### PR TITLE
Fix potential shared memory corruption in resource group slot managem…

### PR DESCRIPTION
**Pull request by paul-guo-**
…ent when query involves entrydb.

We've observed that when we terminate a query that involves entrydb, if QD
detaches from a resource group before the entrydb backend does that in
UnassignResGroup(), data corruption on shared slot pool could happen.  In a
cassert enabled gpdb version, you could see below FATAL message as below when
terminating the query using pg_terminate_backend().

FATAL:  Unexpected internal error (resgroup.c:1545)
DETAIL:  FailedAssertion("!(slot->nProcs == 0)", File: "resgroup.c", Line: 1545)
HINT:  Process 9903 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.

While on producd release that without cassert enabled some subsequent queries
could lead to panic.

We fix this by let the final process (either QD or entrydb process) do clean up
when the nProcs reference number is zero. This is more robust and is less bug
prone cross future code change and upstream merge.

This patch refactors the related resource code a bit and also moves the fault
inject related negative tests ahead so that we could capture some potential
errors which happen later (typically the case in this patch).

Besides, it fixes another panic caused by potential MySessionState NULL
resetting (see code comment for details). That was revealed when running tests
in a dev machine so it's a potential flaky testing part.

Co-authored-by: Paul Guo <pguo@pivotal.io>
Co-authored-by: Ning Yu <nyu@pivotal.io>
Co-authored-by: Gang Xiong <gxiong@pivotal.io>

### CODE REVIEW

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -2313,17 +2318,20 @@ static void
...
-	Assert(Gp_role == GP_ROLE_DISPATCH);
 	Assert(!selfIsAssigned());
 
 	groupPutSlot(group, slot);
 
 	/*
-	 * My slot is put back, then how many queuing queries should I wake up?
-	 * Maybe zero, maybe one, maybe more, depends on how the resgroup's
-	 * configuration were changed during our execution.
+	 * We should wak eup other pending queries on master nodes.
```
Comment (kainwen) [inline src/backend/utils/resgroup/resgroup.c]: wrong space.

Comment (paul-guo-) [reply src/backend/utils/resgroup/resgroup.c]: Fixed. Thanks.

**Comment by asimrp** [comment]
Does this need to be back ported to 6X?

**Comment by paul-guo-** [comment]
> Does this need to be back ported to 6X?

Yes, both 6X and 5X.

**Pull request merged**